### PR TITLE
Set PodStatus QOSClass field

### LINKED ISSUE #33255: Read-only pod status attribute to advertise computed QoS class.
As described by @derekwaynecarr from comments in #32468:

> Developers that build web console experiences need to reproduce our QOS calculation logic. This makes them vulnerable to the same bugs we just fixed in #33108 for each experience they produce. I am supportive of a read-only attribute on PodStatus that states the QOS class of a Pod since once we have qos enforced on the node, we have said a pod cannot change its QOS in its life-time.

@jwforres @vishh @balajismaniam

### PR DESCRIPTION
**Pull request by sjenning**
This PR continues the work for https://github.com/kubernetes/kubernetes/pull/37968

It converts all local usage of the `qos` package class types to the new API level types (first commit) and sets the pod status QOSClass field in the at pod creation time on the API server in `PrepareForCreate` and in the kubelet in the pod status update path (second commit).  This way the pod QOS class is set even if the pod isn't scheduled yet.

Fixes #33255

@ConnorDoyle @derekwaynecarr @vishh

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -1076,6 +1077,7 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontainer.PodStatus) *v1.PodStatus {
 	var apiPodStatus v1.PodStatus
 	apiPodStatus.PodIP = podStatus.IP
+	apiPodStatus.QOSClass = qos.GetPodQOS(pod)
```
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_pods.go]: just add a note that this is needed to ensure upgrade from 1.5 to 1.6 to set the field on upgrade.

**Review by derekwaynecarr — COMMENTED** [review]
one nit, then i will tag.

**Comment by sjenning** [comment]
@derekwaynecarr @vishh rebased and added comment about why we also set the field in the kubelet

File: pkg/registry/core/pod/strategy.go [hunk 2]
```diff
@@ -60,7 +61,8 @@ func (podStrategy) NamespaceScoped() bool {
 func (podStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
 	pod := obj.(*api.Pod)
 	pod.Status = api.PodStatus{
-		Phase: api.PodPending,
+		Phase:    api.PodPending,
+		QOSClass: qos.InternalGetPodQOS(pod),
```
Comment (derekwaynecarr) [inline pkg/registry/core/pod/strategy.go]: i think we should have a test case for this in strategy_test.go

it also would be good if we had a e2e that verified qos class is set as well for each qos tier.  maybe in e2e pods.go?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
thought a little bit more, and think we should have a unit test and e2e test before merging.

File: pkg/registry/core/pod/strategy.go (hunk 2 above)
Comment (vishh) [inline pkg/registry/core/pod/strategy.go]: +1

**Review by vishh — APPROVED** [review]
LGTM for me. Derek's comments have to be addressed though.

File: pkg/registry/core/pod/strategy.go (hunk 2 above)
Comment (sjenning) [inline pkg/registry/core/pod/strategy.go]: I'm pretty sure the unit test would basically be the same as `TestGetPodQOS()` in `qos_test.go`.  I can work on an e2e test.

**Review by derekwaynecarr — APPROVED** [review]

**Comment by derekwaynecarr** [comment]
assuming bazel and all are happy, this is LGTM

**Comment by rmmh** [comment]
@k8s-bot bazel test this

**Comment by sjenning** [comment]
@k8s-bot kops aws e2e test this

**Pull request merged**